### ISSUE TITLE
New extension for linking fences with OS events/file handles

### ISSUE THREAD

**Issue by ratchetfreak** [KhronosGroup/Vulkan-Docs#310]
Currently the only way to wait on the device for a Fence is vkWaitForFences.

However all of this will block the current thread and prevent other OS events (like IO a disk read finishing up, a key press from the user, ...) being handled.

To let other OS events get through to the thread you would need to associate the fence with a Kernel object that can be waited on in bulk (like WairForMultipleObjects or select/epoll).

This would for example be with a  `vkLinkFenceToWin32EventHandleEXT(vkDevice, vkFence, HANDLE);` where if the fence gets signaled the associated event also gets signaled. (only 1 event per fence and vice versa)

Or instead the notification could be done with a IO completion queue:

```C
struct vkLinkFenceToWin32CompletionQueueInfoEXT{
    VkStructureType  sType;
    void* pNext;
    HANDLE completionQueue;
    ULONG_PTR completionKey; 
    OVERLAPPED *overlapped; 
}

vkLinkFenceToWin32CompletionQueueEXT(vkDevice, vkFence, vkLinkFenceToWin32CompletionQueueInfoEXT*);
//or as part of the pNext in fence create info
```

The completionKey and OVERLAPPED pointer are passed back to [GetQueuedCompletionStatus](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364986(v=vs.85).aspx) after the fence is signaled. The OVERLAPPED struct is not copied (to mimic the rest of the win32 api concerning the struct which allows the user to put data after the struct).

I'm not familiar enough with linux programming to judge what the best method for linking fences to file handles is; handle per file or single file acting as a queue? I'll leave that up to others to judge.

**Comment by ratchetfreak** [comment]
With the external_semaphore_* extensions released recently this issue can be resolved by allowing the application to wait on the fd/Handle. Perhaps by adding a flag to let the application signal that it wants to do that.

**Comment by cubanismo** [comment]
Methods to do this with fences are under discussion.  However, waiting on semaphores from the CPU is not supported.  They are intended for GPU-side waits.

**Comment by daiheitan** [comment]
1.0.54.0 adds `vkGetFenceFdKHR`, I guess it's the way to go?

**Comment by ratchetfreak** [comment]
@daiheitan looks like however it's still not documented allow use with WaitForMultipleObjects or select/epoll

**Comment by daiheitan** [comment]
@ratchetfreak Yeah... Seems the latest Nvidia driver has this extension, but with no select/epoll support. I'm still testing the API.

**Comment by timofonic** [comment]
@ratchetfreak @oddhack @cubanismo @daiheitan Any news?

**Comment by Ralith** [comment]
Any news on this? Would be very convenient.

**Comment by expipiplus1** [comment]
VK_KHR_external_fence isn't really intended for this use case. As far as I understand it, it's intended to be used for using fences across multiple instances/devices.

Having said that, it would be very convenient to extend VkExternalFenceHandleTypeFlagBits with an entry requiring support for epoll to facilitate application usage of these fds. 

(perhaps best as a separate issue: does anyone know what a "Linux sync file" is (as mentioned in the docs for VkExternalFenceHandleTypeFlagBits)?)

**Comment by expipiplus1** [comment]
Was this ever resolved inside Khronos? It would be handy to use the OS's waiting primitives instead of Vulkan's.

I noticed recently that @critsec suggests here (https://stackoverflow.com/questions/42721870/means-for-calling-back-from-vulkan-to-host-application) that this is possible with the `FENCE_FD` handle type did you mean `SYNC_FD`?

**Comment by critsec** [comment]
Getting a `SYNC_FD` handle type using `vkGetFenceFdKHR` on Linux/Android will give you a fd that is compatible with epoll. This is definitely a use case that VK_KHR_external_fence is intended for.

I don't think there's an equivalent for Windows yet; the `OPAQUE_WIN32` and `OPAQUE_WIN32_KMT` handle types aren't compatible with WaitForMultipleObjects.

A fallback is to use a separate utility thread that waits using `vkWaitForFences` and then signals something that is compatible with what you want to do. This is clearly less than ideal (extra thread scheduling delays, etc.) but works well enough in some cases.

**Comment by cubanismo** [comment]
Yeah, the FDs you pull out of fences on the NV Linux driver aren't going to do much if you poll() them. If we did have an API to get the right FD, because of the way our HW/driver manages fences (potentially other drivers/HW as well, not sure), the usage would have to be something like:

```
VkFence fence = createAndSignalSomeFence();
vkGetFenceFdKHR(..., fence, ... handleType = VK_EXTERNAL_FENCE_HANDLE_TYPE_POLL_FD ..., &fd);

while (true) {
   poll(..., fd, ...);
   // Might wake up spuriously, beyond the usual EINTR reasons. Check actual fence status:
   if (vkGetFenceStatus(dev, fence)) break;
}
// Fence is done.
```

And similar for timeline semaphores.  Another place this gets complicated is the interactions with other external handle types and by implication, WSI.  E.g., if you try to do this with a fence you pass to vkAcquireNextImageKHR(), it may not work as you expect by the time we fold in the "payload" concept.

That said, if there are demonstrated use cases where this would be better than the alternatives (E.g., spin up a waiter thread), we could pursue an extension, but I haven't heard much demand thus far, so I haven't pursued it. Most people seem OK with dedicating a thread to it, or at least happy enough that they haven't made it their highest priority Vulkan spec issue.  If you need it on windows, things are a little more complicated too, but it's probably doable with win32 event objects there.

**Comment by expipiplus1** [comment]
> Getting a `SYNC_FD` handle type using `vkGetFenceFdKHR` on Linux/Android will give you a fd that is compatible with epoll. This is definitely a use case that VK_KHR_external_fence is intended for.

At the least, would it be worth enshrining this in the spec, perhaps under a physical device property?

> Another place this gets complicated is the interactions with other external handle types and by implication, WSI. ...

Thinking about this a little more, perhaps a more general, platform agnostic solution would be to have callbacks executed by Vulkan when waiting is done, something like

```C
void myCallback(void* myData, uint32_t semaphoreCount, VkSemaphore* semaphores) {
  // touch my fd or whatever
}

void myProgram() {
  VkWaitSemaphoreCallbackInfo waitCallbackInfo;
  waitCallbackInfo.sType = VK_STRUCTURE_TYPE_WAIT_SEMAPHORE_CALLBACK_INFO;
  waitCallbackInfo.pNext = NULL;
  waitCallbackInfo.callback = myCallback;
  waitCallbackInfo.data = NULL;

  VkSemaphoreWaitInfo waitInfo;
  waitInfo.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO;
  waitInfo.pNext = &waitCallbackInfo;
  waitInfo.semaphoreCount = 1;
  waitInfo.pSemaphores = &mySem;
  waitInfo.pValues = &myValues;
  const VkResult r = vkWaitSemaphores(dev, waitInfo, 0);

  switch(r) {
    case VK_SUCCESS:
      // myCallback has been called
    case VK_TIMEOUT:
      // myCallback may not have been called yet
  }
}
```

It may not be appropriate to reuse `vkWaitSemaphores` for this, as it's now quite a distinct operation (building up state in the driver). On the other hand `VkSemaphoreWaitInfo` is precisely the type which should go with `VkWaitSemaphoreCallbackInfo`.

This would also neatly push the handling of spurious wakeups back into the driver which already has to deal with it.

> That said, if there are demonstrated use cases where this would be better than the alternatives (E.g., spin up a waiter thread), we could pursue an extension, but I haven't heard much demand thus far, so I haven't pursued it. Most people seem OK with dedicating a thread to it, or at least happy enough that they haven't made it their highest priority Vulkan spec issue.

The desire here: https://github.com/vulkano-rs/vulkano/issues/666 was to integrate Vulkan more cleanly with various Rust job-systems/io-managers (although there's nothing Rust-specific here).

My use case would be to use Vulkan with concurrent Haskell without having my threads robbed having to service waits, the GHC event manager is good at waiting on FDs however.

I suppose that in general this would enable Vulkan to be integrated into event systems without them having to handle anything Vulcan-specific.

Certainly these can both be done with waiter threads but like @critsec says, this is clearly less than ideal.

> If you need it on windows, things are a little more complicated too, but it's probably doable with win32 event objects there.

Not the driver's concern any more using the callback approach :wink:

**Comment by Ralith** [comment]
A callback invoked while a thread is blocked isn't any better than the status quo; you could just wait then call the thread. The appeal of epoll/whatever compatibility is that it allows you to compose waiting for a vulkan resource with other asynchronous operations within a single thread.

Personally I've addressed my needs here by just polling timeline semaphores in various patterns. It's not the most elegant solution, but conversely, it's much more portable than a platform-specific multiplexing scheme.

**Comment by expipiplus1** [comment]
> A callback invoked while a thread is blocked isn't any better than the status quo; you could just wait then call the thread. The appeal of epoll/whatever compatibility is that it allows you to compose waiting for a vulkan resource with other asynchronous operations within a single thread.

One can build this epoll compatibility on top of callbacks though.

```
myFd = eventfd(...)
vkWaitSemaphores(... callback which writes to myFd)
epoll_wait( read myFd and any other async operations )
```

At some point the driver's going to have to call `write` or `SetEvent`, is it any matter if this is done in an application-supplied callback? (I think the only case where this would matter would be if the driver does this from some privileged context)

**Comment by Ralith** [comment]
If your `vkWaitSemaphores` call doesn't block, that could work, but I'm not sure how an ICD could implement that, except by spawning a background thread on its end, which seems a bit silly.

**Comment by expipiplus1** [comment]
Some drivers do have userspace threads kicking around, but in most cases you're probably right. Certainly with epoll/WaitEvents support would allow one to implement this themselves anyway.

For the case of timeline semaphores, `eventfd` objects keep a 64bit value inside them, so that could be used (although the application would wake on every change and have to check themselves).

**Comment by cubanismo** [comment]
> > Getting a SYNC_FD handle type using vkGetFenceFdKHR on Linux/Android will give you a fd that is compatible with epoll. This is definitely a use case that VK_KHR_external_fence is intended for.
> 
> At the least, would it be worth enshrining this in the spec, perhaps under a physical device property?

I believe this is already implied. The spec specifies the properties of FDs returned for SYNC_FD (Basically states it's compatible with either the Android fence FD or Linux sync FD kernel primitives for a given system, IIRC), and I believe Linux itself defines the operations possible on a sync FD.

> Thinking about this a little more, perhaps a more general, platform agnostic solution would be to have callbacks executed by Vulkan when waiting is done

That's just going to mean the driver spinning up that extra thread for you opaquely, without the ability for you to control its affinity, priority, etc. No value-add for doing this in the driver, so not a good candidate for Vulkan functionality. Just spin your own thread as suggested above, use that to make a callback to the rest of your code, signal an event FD, etc.

**Comment by oddhack** [comment]
@ratchetfreak could this be closed, based on the most recent comment from @cubanismo ?

**Comment by msharov** [comment]
> That said, if there are demonstrated use cases where this would be better than the alternatives (E.g., spin up a waiter thread), we could pursue an extension, but I haven't heard much demand thus far, so I haven't pursued it. Most people seem OK with dedicating a thread to it, or at least happy enough that they haven't made it their highest priority Vulkan spec issue.

That's only because most people have a really hard time figuring out what the intended solution is. Having an fd to wait on for fences is better than the waiter thread for pretty much all applications, since pretty much all applications already need to wait on something other than the fences. Any application other than a background compute batch job will need to wait for user input from some fd. Requiring every interactive application to have a separate thread just to wait on vulkan fences seems rather heavy handed.

Unless, that is, you have some other preferred solution on how an interactive application is supposed to wait for fences. If you do, I haven't come across any description of it anywhere. All the examples I've seen wait for fences after submitting the command buffer for the frame, resulting in the application being responsive only in the very short time between vsync and the submission of the next frame. Definitely not user friendly.

**Comment by Ralith** [comment]
@msharov, have you tried using a sync FD? Seems like that's supported.

**Comment by msharov** [comment]
> @msharov, have you tried using a sync FD? Seems like that's supported.

This issue is about whether vkGetFenceFdKHR returns a sync fd that can be polled. The last comments seem to indicate that it does, so that's what I will try using. It would be nice if there was example code for that, since it is the usual scenario for any UI application and vulkan always requires a mountain of non-obvious code to do anything.

I also disagree with @cubanismo's comment that the spec is already sufficiently clear on the subject. The ability to poll the returned sync fd should be explicitly mentioned, because it is such a common need and it is not at all obvious to people who do not routinely live in the guts of the linux kernel, where the concept of "sync file" comes from.

**Comment by msharov** [comment]
I have tried getting a sync fd with vkGetFenceFdKHR and it does not work.

I am currently modifying the code from https://vulkan-tutorial.com/, and have tried the following. Create the device with the VK_KHR_EXTERNAL_FENCE_FD_EXTENSION_NAME enabled. Create the fence with the VkExportFenceCreateInfo:

```c++
VkFenceCreateInfo info{};
info.sType = VK_STRUCTURE_TYPE_FENCE_CREATE_INFO;

VkExportFenceCreateInfo exinfo{};
exinfo.sType = VK_STRUCTURE_TYPE_EXPORT_FENCE_CREATE_INFO;
exinfo.handleTypes = VK_EXTERNAL_FENCE_HANDLE_TYPE_OPAQUE_FD_BIT| VK_EXTERNAL_FENCE_HANDLE_TYPE_SYNC_FD_BIT;

info.pNext = &exinfo;

vkCreateFence (device(), &info, nullptr, &_f);
```
Then in drawFrame I do:
```c++
int ffd = -1;
auto pgffd = (PFN_vkGetFenceFdKHR) vkGetInstanceProcAddr (_instance, "vkGetFenceFdKHR");

VkFenceGetFdInfoKHR fgfi {};
fgfi.sType = VK_STRUCTURE_TYPE_FENCE_GET_FD_INFO_KHR;
fgfi.fence = frame_fence;
fgfi.handleType = VK_EXTERNAL_FENCE_HANDLE_TYPE_SYNC_FD_BIT;
result = pgffd (_device, &fgfi, &ffd);
if (result != VK_SUCCESS) {
    printf ("could not get fence fd: %d\n", result);
    exit (EXIT_FAILURE);
}

struct pollfd pfd {};
pfd.fd = ffd;
pfd.events = POLLIN;
if (0 > poll (&pfd, 1, 1000))
    perror ("poll");
close (ffd);
```
vkGetFenceFdKHR then fails with -3, `VK_ERROR_INITIALIZATION_FAILED`.

If I do not include `VK_EXTERNAL_FENCE_HANDLE_TYPE_OPAQUE_FD_BIT` during fence creation, then the failure is -8, `VK_ERROR_FEATURE_NOT_PRESENT`.

If I ask for `VK_EXTERNAL_FENCE_HANDLE_TYPE_OPAQUE_FD_BIT` instead of sync, I do get a valid fd and can call poll. poll then times out after one second, so likely the fd type is not suitable.

So where is the solution here? Did I do something wrong here? Or is it the spec or the driver?

I'm using linux kernel 6.0.9-arch1-1, nvidia 520.56.06-11.

**Comment by chloekek** [comment]
I’ve tried it on Linux about a year ago and likewise it didn’t work. With AMD discrete obtaining, the file descriptor resulted in device lost. With Intel integrated, poll would block indefinitely. I’ll see if I can recover the code and try again.

**Comment by msharov** [comment]
> I agree. However, there are a lot of things Vulkan needs, and there's been little noise made about this issue over the years as compared to others. If there are more users running into this, pile on the issue with comments/+1s on existing comments, and it should get more attention.

Counting comments here is not a good metric for what is or what is not needed. Most people would never come to this project intentionally; I only came while searching for some solution to polling fences. Others will likely give up much sooner and abandon trying to use Vulkan altogether. I mean, integration with the windowing system is some really basic stuff. If you are not getting this right, it is tempting to just declare the whole thing "not ready for general use".

> It's not the business of the Vulkan spec to define the properties of things that, as you imply and I agree, are defined by the guts of the Linux kernel.

It is the business of the Vulkan spec to define functionality that every application using Vulkan will need. Since every non-trivial application requires reading user input, you should support a simple way of doing it. Requiring the use of an accidentally supported hack is just unprofessional.

> The spec makes minimal claims as to the Vulkan-external functionality provided by the FDs returned for the OPAQUE_FD handle types because there is nothing else to define the functionality of those types. They are an invention of Vulkan, to the extent they're novel at all.

So what is the poing of all these external resource extensions then? I can't get a working fence fd, I can't import a non-vulkan object like an X11 pixmap (yet another absent essential feature required to implement a compositing window manager); so what can I do with it and why did you bother implementing them in the first place?

> 2. That SYNC_FD is supported at all on the driver you're using? E.g., currently the NVIDIA driver only supports SYNC_FD on Tegra, but support is coming on desktop.

Sigh. I guess Vulkan just isn't ready for general use. I'll stop pestering you now.

**Comment by cubanismo** [comment]
> It is the business of the Vulkan spec to define functionality that every application using Vulkan will need.

No, it's really not. I can't stress that enough. Even the way that sentence is worded implies such a thing is beyond the scope of the Vulkan spec. The Vulkan spec defines Vulkan, not things applications using Vulkan will need. Sometimes hints are given as non-normative "notes," but I don't think this particular bit of info even warrants one of those. Also, many applications have thus far managed to use Vulkan and accept user input without the proposed functionality. I understand what you're getting at (A single-poll-call event loop design), but it's deceptive to imply it's the only valid or even optimal way to write an interactive Vulkan application on a POSIX or POSIX-like system.

> Since every non-trivial application requires reading user input, you should support a simple way of doing it. Requiring the use of an accidentally supported hack is just unprofessional.

Are you calling support for poll(sync_file) an accident and a hack? It's neither. Neither is the option of spinning up a thread to signal your main poll()/epoll() call. That's how we assumed an application would accomplish this given the current spec. Threads are great for this sort of thing, and they have a smaller OS-specific Vulkan API surface (I.e., none) than the proposed pollable FD functionality, though as noted, I personally agree a pollable opaque FD extension would still be generally useful enough to include in the API. That said, have you tried implementing this using a helper thread? I won't claim it's minimal, but it is simple unless there are some special circumstances.

> I can't import a non-vulkan object like an X11 pixmap

I'd be happy to work on this, but I haven't found any other driver vendors who care to collaborate on features for X11 anymore, so it would end up an NVIDIA-only extension at this point, which isn't that useful. I agree Vulkan is not a good choice for an X11 composite manager right now, but that's pretty far from making Vulkan or even the external resource extensions "not ready for general use."

> I'll stop pestering you now.

If you or anyone else has constructive comments, happy to hear them. I understand it is frustrating when an API is lacking a desired feature. Note it is possible to add extensions to the Vulkan specification by making a pull request against this repository, and you can implement those extensions in various open source Vulkan drivers maintained as part of the Mesa project. Patches always get more attention than requests for help, and specifications implemented in open source drivers have historically been a great way to get the attention of the closed source driver maintainers. I'm not saying users who have bought a graphics card or SoC *should* have to go through such lengths to get a feature in, but it's an option.

**Comment by msharov** [comment]
> I understand what you're getting at (A single-poll-call event loop design), but it's deceptive to imply it's the only valid or even optimal way to write an interactive Vulkan application on a POSIX or POSIX-like system.

Every X11 framework is built on a single-thread event loop design, because that's what Xlib is written for. Multithreading is not for the faint of heart and certainly not for newbies. I avoid it at all cost, even though I actually do have enough experience to do it properly. Event loops are better than multithreading in every way. They are responsive, simple to debug, and have the same low-level efficency that Vulkan has over OpenGL. I think it is not at all unreasonable to call it the optimal way to write everything.

> Neither is the option of spinning up a thread to signal your main poll()/epoll() call. That's how we assumed an application would accomplish this given the current spec. Threads are great for this sort of thing

Threads are not great for this sort of thing. It only seems that way because threads are implemented in the kernel and in the CPU, and you do not normally see just how much hassle and performance impact they bring.

Another way of looking at it is that a thread is an unnecessary abstraction here. The Vulkan driver is not going to sit around in a spinlock, waiting for the GPU to finish rendering. It gets an interrupt (or something, I'm fuzzy on the details) and returns control to the application. Why enforce this wait when you could have just passed the interrupt via some event fd? No threads required; your internal design is *already* event driven. Forcing threading *is* a hack, and an ugly one.

> That said, have you tried implementing this using a helper thread?

Obviously, it is my only option.

> I'd be happy to work on this, but I haven't found any other driver vendors who care to collaborate on features for X11 anymore

Yeah. Wayland was a pile of junk a decade ago and it still is. You'd think people would create something usable to switch to before deprecating existing software that works just fine...

> If you or anyone else has constructive comments, happy to hear them. I understand it is frustrating when an API is lacking a desired feature. Note it is possible to add extensions to the Vulkan specification by making a pull request against this repository, and you can implement those extensions in various open source Vulkan drivers maintained as part of the Mesa project.

Implementing graphics driver guts is way beyond my level of expertise. I have no experience in writing drivers and have little interest in gaining any.

Yes, I know, I don't like what is out there, so I should write my own. And I have been. There's just so much of it to write that it gets frustrating that after many years of writing code I still have hardly anything that works.

**Comment by smumriak** [comment]
> That said, have you tried implementing this using a helper thread? I won't claim it's minimal, but it is simple unless there are some special circumstances.

I have and it **is** pretty simple. I literally have a RunLoop built atop of timeline semaphores that is running in background thread: https://github.com/smumriak/AppKid/blob/main/Volcano/Sources/Volcano/Utilities/Synchronization/SemaphoreRunLoop.swift

I've started doing the same for fences, but right now I have no use for that so it's incomplete: https://github.com/smumriak/AppKid/blob/main/Volcano/Sources/Volcano/Utilities/Synchronization/FenceRunLoop.swift

**Comment by buzmeg** [comment]
Languages with a managed runtime often already have a way to deal with FDs that block.  They can do things like release global locks, fire off GC, etc. during such times.

Forcing those runtimes to spawn a separate OS-managed thread just to handle blocking calls throws away all of that hard won integration.

**Comment by herabit** [comment]
This functionality would be ***extremely*** handy for situations like anything using Vulkan for video processing that has to interact with the network.

Having something that'd work well with epoll-compatible asynchronous runtimes would be a godsend.

I'm in the midst of building a screensharing application that will be re-encoding the video stream as needed before sending the stream to clients. I plan on, at least on Linux, only supporting PipeWire, and I'm currently designing and implementing a PipeWire library that'll work with tokio, and eventually other Rust async runtimes.

How I'm getting my frames is already asynchronous, I feel it should be trivial to integrate the asynchrony of Vulkan into already existing async runtimes, such as those in Rust.

The two approaches for Rust, at least when dealing with async Rust, is to operate a spin loop that polls a fence, if the fence is finished, then returns a `Poll::Ready(..)` value, and if it isn't complete, instead tell the asynchronous runtime to poll the future, again, at a later point.

Functionally it's a spinloop, which isn't really, well, ideal. Even if this spinloop yields control back to the scheduler for the async runtime, it's still scheduling itself to be polled, again, and again, and again.

You can see this approach in Vulkano.

The alternative is, well to spin up a thread, block, then notify the async runtime when it completes. This avoids the issues of a spinloop, but ultimately still incurs the cost of needing another thread, and needing to notify a different thread.

If epoll could be integrated, at least on Linux, I'd be able to use what tokio is already using to concurrently execute IO, without *needing* another thread, and without *needing* to slam the scheduler each time polling fails.

Are there any GPU drivers that support this functionality at least unofficially? Because if so, I would like to be able to support this capability for when I'm using those drivers, and otherwise defer to the less-than-ideal solutions.

For I/O bound applications that want to use Vulkan, this would permit much more versatile usage patterns.

In a world where Vulkan Video Coding exists, this would be beautiful.
